### DISCUSSION TITLE
Diagnosing "White Noise" in PyCaret

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1800] [FAQs]
# Background

White noise is time series data that does not have any signal that can be modeled i.e. it only has noise. Diagnosing white noise  is an important part of exploratory analysis and can save modelers a lot of time by avoiding fitting more expensive time consuming models. Plus modeling this data with anything other than a `naive` Grand Means model would be an inappropriate way to capture the underlying process information (more on this later).

## Diagnostics using ACF

There are several tools and checks available to test for white noise. For example, we can use ACF plots to look for significant lags. Data where the ACF magnitude is not significant at any lag is indicative of white noise. In other words, there is no temporal signal that can be derived from any lagged values of the dataset. More information about this can be found in **[this FAQ](https://github.com/pycaret/pycaret/discussions/1765)**

## Diagnostics using `Ljung-Box` test

Another tool available to modelers is a statistical test called **`Ljung-Box`** test. This will test all the lagged values of the ACF simultaneously (as a group rather than individually as it would be done when visual inspecting ACF plots) to determine if they are consistent with white noise. The null and alternate hypothesis are as follows

**H<sub>0</sub>: &rho;<sub>1</sub> = &rho;<sub>2</sub> = &rho;<sub>3</sub> = ... = &rho;<sub>K</sub> = 0**
**H<sub>a</sub>: At least 1 &rho;<sub>k</sub> &ne; 0 for 1 &le; k &le; K**

In simpler words

**H<sub>0</sub>: Data is consistent with white noise**
**H<sub>a</sub>: Data is inconsistent with white noise**

It is often advisable to test the data set for 2 different values of K (24 and 48 are often used as reasonable defaults).

# Diagnosing White Noise in `pycaret`

It is best to model this kind of time series data set with a `naive` (also known as Grand Means) model where the prediction is simply the mean value of the data used to train the model. This is the "best" estimate of any future prediction when the data is white noise. Hence, it is critical to make sure we check for white noise before we start modeling any time series data set. Let's see how we can do this in `pycaret`. Note that this FAQ is also available as a **[notebook here](https://gist.github.com/ngupta23/b63c7480e3eb63dc30ebac483177f4d1)** so you can run the code yourself.

```python
#### Install library (one time) ----
!pip install pycaret-ts-alpha -U

#### Import libraries ----
from pycaret.datasets import get_data
from pycaret.internal.pycaret_experiment import TimeSeriesExperiment

#### Get the data ---
y = get_data("1", folder="time_series/white_noise")

#### Setup the experiment ----
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=30, seasonal_period=1, session_id=42)
```

![image](https://user-images.githubusercontent.com/33585645/140304190-64a1dc6b-668c-4840-9a1e-4ef7d5a89520.png)

When setup is run, some exploratory analysis is automatically done in the background (e.g. `Ljung-Box` test is run for K=24 and 48 by default) and the results are presented for the modeler. In this case, `pycaret` diagnosed this as white noise data.

In addition, modelers can perform their own exploratory analysis using the interactive EDA capability available in `pycaret`

```python
#### Plot the dataset & its ACF ----
exp.plot_model(plot="diagnostics")
```
![image](https://user-images.githubusercontent.com/33585645/140304504-1bb5dfaf-d15b-4bd0-a7ba-8aa79de2779e.png)

Again, from the diagnostic plots, we can see that the histogram, QQ plot and ACF all seem to indicate that the data is consistent with white noise with a mean of ~ 176.

We noted above that Ljung-Box test was performed during the `setup` phase. Modelers can get the details of this test using the following

```python
#### Without an estimator, the test is applied to the original dataset ----
exp.check_stats(test="white_noise")
```

![image](https://user-images.githubusercontent.com/33585645/140304985-b990de63-4642-468e-b2f0-95f6d9612b77.png)

As we can see, at a significance level &alpha; = 0.05, we fail to reject the null hypothesis for both K=24 and K=48 since the p-value is greater than 0.05.

All this exploratory analysis indicates that the data is consistent with White Noise! That's it for this FAQ. In a subsequent FAQ, we will see how we can model white noise after we have diagnosed it. 

Happy forecasting!

# Suggested Next Reads

- https://github.com/pycaret/pycaret/discussions/1837

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
